### PR TITLE
Backend for scenario and execution views

### LINKED ISSUE #113: Web: Implement backend for Scenario and Execution views
The backend should handle all API calls from `web/src/api.tsx` by utilizing the DB and the Execution model.

### PR DESCRIPTION
**Pull request by 7erra**
This PR implements the backend for the scenario and execution views. Instead of using dummy data the data is now pulled from the database and the active_executions variable.

Closes #113

### CODE REVIEW

File: server/models.py [hunk 1]
```diff
@@ -16,18 +17,25 @@
 class Scenario(db.Model):
     id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
     name: Mapped[str] = mapped_column(nullable=False)
+    executions: Mapped[List["Execution"]] = relationship(
+        back_populates="scenario")
```
Comment (simohlsen) [inline server/models.py:21]: Why is this needed? Is this also a column in the DB or just a reference in the ORM-object?

File: server/models.py [hunk 2]
```diff
@@ -16,18 +17,25 @@
...
 
 
 class Execution(db.Model):
     id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
-    scenario_id: Mapped[int] = mapped_column(ForeignKey("scenario.id"), nullable=False)
+    scenario_id: Mapped[int] = mapped_column(
+        ForeignKey("scenario.id"), nullable=False)
+
+    scenario: Mapped["Scenario"] = relationship(back_populates="executions")
+    players: Mapped[List["Player"]] = relationship(back_populates="execution")
```
Comment (simohlsen) [inline server/models.py:30]: Are these also a columns in the DB or just references in the ORM-object?

File: server/models.py [hunk 3]
```diff
@@ -42,32 +50,40 @@ class Player(db.Model):
...
     )
-    location_id: Mapped[int] = mapped_column(ForeignKey("location.id"), nullable=False)
+    location_id: Mapped[int] = mapped_column(
+        ForeignKey("location.id"), nullable=False)
     role_id: Mapped[int] = mapped_column(ForeignKey("role.id"), nullable=False)
     alerted: Mapped[bool] = mapped_column(nullable=False)
     activation_delay_sec: Mapped[int] = mapped_column(nullable=False)
 
+    execution: Mapped["Execution"] = relationship(back_populates="players")
+
```
Comment (simohlsen) [inline server/models.py:60]: Why is this needed? Is this also a column in the DB or just a reference in the ORM-object?

File: server/web/api/register.py [hunk 4]
```diff
@@ -1,27 +1,55 @@
...
 from app_config import csrf
-from utils.tans import uniques
-from .test_data import test_execution
+from execution.run import activate_execution, active_executions
 
 # FIXME sollten wir zusammen mit dem Web package iwann mal umbenennen
 api = Blueprint("api-web", __name__)
 
 
+class ExecutionIdNotFound(Exception):
```
Comment (YIllmann) [inline server/web/api/register.py:19]: If you want to create more of such exceptions, we might relocate them into an extra package. For now it is ok :).

File: server/web/api/register.py [hunk 5]
```diff
@@ -69,58 +97,57 @@ def login():
...
 
 
 @api.post("/execution/<int:id>/start")
 @admin_only
 def start_execution(id: int):
-    test_execution["status"] = "Aktiv"
-    return test_execution
+    execution = try_get_execution(id)
+    execution.status = Execution.Status.RUNNING
+    return Response(status=204)
```
Comment (YIllmann) [inline server/web/api/register.py:132]: why no content instead of 200? Based on the state change i think 200 fits better.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

**Review by simohlsen — CHANGES_REQUESTED** [review]

File: web/src/api.tsx [hunk 6]
```diff
@@ -84,8 +87,8 @@ export async function getExecutionStatus(id: string): Promise<object> {
 export async function startExecution(
   id: string,
   formData: FormData,
-): Promise<object> {
-  return tryFetchJson(`execution/${id}/start`, {
+): Promise<Response> {
+  return tryFetchApi(`execution/${id}/start`, {
```
Comment (YIllmann) [inline web/src/api.tsx:91]: For the mobile api we decided not to put any request parameter to the url to keep the url clean.
I would suggest you use the same standard and use query parameters instead of path variables, especially for POST request.

**Comment by simohlsen** [comment]
I would suggest to increase the line break limit to 120 because most of the breaks added don't really improve readability.

File: server/web/api/register.py [hunk 7]
```diff
@@ -69,58 +97,57 @@ def login():
...
+    execution = try_get_execution(id)
+    execution.status = Execution.Status.FINISHED
+    return Response(status=204)
+
+
+@api.post("/execution/<int:id>/player/<tan>/status")
+@admin_only
+def change_player_status(id: int, tan: str):
+    try:
+        status: bool = request.form["alerted"] == "1"
```
Comment (YIllmann) [inline server/web/api/register.py:147]: if u send `"True"` or `"False"` in the form, you can use bool(request.form["alerted"]). Then you dont have to hardcode your expected value.

**Comment by 7erra** [comment]
It certainly increases the readability in my editor. With a line width of 80 I can have two files open side by side.

**Review by YIllmann — CHANGES_REQUESTED** [review]

File: server/web/api/register.py (hunk 5 above)
Comment (YIllmann) [reply server/web/api/register.py]: That is because I implemented them ^^. Overall the status indicates a valid request but no changes in the backend due to missing content. 
- app: The methods blocks every run request from any mobile device to prevent game manipulation before the action has started. 
- lobby: "The execution has not been started yet" - same reason here.
- test_lobby is a test case for the lobby implementation and has no further impact on the app.

File: server/web/api/register.py (hunk 7 above)
Comment (YIllmann) [reply server/web/api/register.py:147]: just realized my suiggestion is invalid. consider it resolved.

File: server/web/api/register.py [hunk 8]
```diff
@@ -69,58 +97,57 @@ def login():
...
-        {"id": 10023, "name": "Busunfall", "players": 5},
-        {"id": 900323, "name": "Explosion im Wohnviertel", "players": 10},
-    ]
+    return [{"id": scenario.id, "name": scenario.name, "executions": [execution.id for execution in scenario.executions]}
+            for scenario in models.Scenario.query]
 
 
 @api.post("/scenario/start")
-@admin_only
+# @admin_only
```
Comment (H1ghBre4k3r) [inline server/web/api/register.py:105]: Commented code is bad 👻 (or should be removed before merging)

File: web/src/components/CsrfForm.tsx [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { Form, FormProps } from "react-router-dom"
+import { CsrfInput } from "../contexts/csrf"
+import { ReactElement, ReactNode } from "react"
+
+export default function CsrfForm({
+  children,
+  ...props
+}: FormProps & { children?: ReactNode }): ReactElement {
```
Comment (H1ghBre4k3r) [inline web/src/components/CsrfForm.tsx:8]: For this type you can just use `PropsWithChildren` provided by React.

File: web/src/routes/execution.tsx [hunk 10]
```diff
@@ -27,46 +36,70 @@ function TanCard({ tan }: { tan: string }): ReactElement {
...
 
 export default function Execution(): ReactElement {
   const loaderData = useLoaderData()
-  const actionData = useActionData()
+
   useEffect(() => {
-    if (isExecutionData(actionData)) {
-      setExecution(actionData)
+    if (isExecutionData(loaderData)) {
+      setExecution(loaderData)
```
Comment (H1ghBre4k3r) [inline web/src/routes/execution.tsx:93]: I would prefer to locate this code below the definition of `execution` and `setExecution`.

**Review by H1ghBre4k3r — CHANGES_REQUESTED** [review]

File: server/web/api/register.py (hunk 8 above)
Comment (7erra) [reply server/web/api/register.py]: I forgot to reenable that after testing. Should be fine now.

File: web/src/api.tsx (hunk 6 above)
Comment (H1ghBre4k3r) [reply web/src/api.tsx]: Since we already decided on that, we should probably stick to that. However, I am not a fan of doing stuff like that via query parameters. 

Usually, in REST apis, the path is used to identify a resource and the query is used to, for example, provide filter criteria (e.g., in a search).

File: server/web/api/register.py (hunk 7 above)
Comment (7erra) [reply server/web/api/register.py]: I have implemented an equivalent way by using [python-string-util's booleanize](https://python-string-utils.readthedocs.io/en/develop/manipulation.html?highlight=booleanize#string_utils.manipulation.booleanize)

File: web/src/api.tsx (hunk 6 above)
Comment (H1ghBre4k3r) [reply web/src/api.tsx]: Generally, I think we should stick more to the semantics of REST. 

I dont really see the point of having an extra "start" endpoint - we should use a simple `POST` request against `/execution/{id}`. As a counterpart, `DELETE /execution/{id}` should stop the execution. 

Creating an execution should be `POST /execution`, while `GET /execution` retrieves all executions and `GET /execution/{id}` retrieves the execution with the provided id.

File: server/web/api/register.py [hunk 11]
```diff
@@ -69,58 +66,59 @@ def login():
...
+        return active_executions[id].to_dict()
 
+    # Failure
+    raise NotFound(f"Execution with id={id} does not exist")
 
-@api.get("/execution/<int:id>")
-@admin_only
+
+@api.get("/execution")
+@required("id", int, RequiredValueSource.ARGS)
```
[the remark this answers is not in the document]
Comment (7erra) [reply server/web/api/register.py:85]: So do I but I suggest leaving that for another PR. In that future PR the backend for the app could also be adjusted. Right now we follow the same standard for web and app backend at least.

File: server/web/api/register.py [hunk 12]
```diff
@@ -69,58 +66,51 @@ def login():
...
+@required("running", booleanize, RequiredValueSource.FORM)
 @admin_only
-def start_execution(id: int):
-    test_execution["status"] = "Aktiv"
-    return test_execution
+def change_execution_status(id: int, running: bool):
+    execution = try_get_execution(id)
+    execution.status = Execution.Status.RUNNING if not running else Execution.Status.PENDING
+    return Response(status=200)
```
Comment (simohlsen) [inline server/web/api/register.py:100]: Isn't `Execution.Status.FINISHED` missing here?

File: web/src/api.tsx (hunk 6 above)
Comment (YIllmann) [reply web/src/api.tsx]: "I don't remember agreeing on anything and would therefore advocate for using path parameters." 
This is a very critical call to make. Maybe you missed the discussion due to your illness or less interests in the discussion because we were discussing the interface to the mobile api and was therefore less relevant. I was just pointing out the facts we have already implemented so far.

I see the Clean URL reference and will revoke my suggestion. However i do not agree to use POST execution as start endpoint, as well as DELETE to end execution. This leads to further confusion in the semantics and the current willingnes to document the pushed code supports that statement.

**Review by YIllmann — APPROVED** [review]

File: server/web/api/register.py (hunk 12 above)
Comment (7erra) [reply server/web/api/register.py]: I have changed the button to a dropdown to make it possible to set the status to anything you want (except unknown).

Comment (simohlsen) [reply server/web/api/register.py]: Wouldn't it be nicer, if there is a start button for pending executions and the start button turns into a stop button when the execution is running? Just a suggestion.

Comment (H1ghBre4k3r) [reply server/web/api/register.py]: @simohlsen I totally agree with you. However, this should probably be part of a future PR.

File: web/src/api.tsx (hunk 6 above)
Comment (H1ghBre4k3r) [reply web/src/api.tsx]: @YIllmann I think, we should discuss the semantics of our API at a later point and maybe change the endpoints in a future PR. For now, this is fine for me.

**Review by H1ghBre4k3r — APPROVED** [review]

**Comment by H1ghBre4k3r** [comment]
LGTM. However, there is a conflict regarding `server/models.py`.
